.github/scripts/platform_lib.py

```diff
@@ -0,0 +1,377 @@
+    
+    def terminate_instances(self, gh_token: str, tag_value: str) -> None:
+        """ Stops the instances specified by 'tag_value' """
+        self.change_workflow_instance_states(gh_token, tag_value, 'terminate')
+    
+    def get_manager_hostname(self, tag_value: str) -> str:
+        """ Returns the hostname of the ci manager specified """
+        return f"centos@{self.get_manager_ip(tag_value)}"
+
+class AWSPlatformLib(PlatformLib):
+
+    def __init__(self, deregister_runners: Callable[[str, str], None]):
+        if os.path.exists(os.path.expanduser('~/.aws/config')): # only set client if this exists
+            self.client = boto3.client('ec2')
+        self.manager_filter = {'Name': 'tag:ci_manager', 'Values' : ['']}
+        self.deregister_runners = deregister_runners
+    
+    def get_filter(self, tag_value: str) -> Dict[str, Any]:
+        return {'Name': 'tag:' + unique_tag_key, 'Values' : [tag_value]}
+
+    def get_manager_tag_dict(self, sha, tag_value):
+        """ Populates a set of tags for the manager of our CI run """
+        # Note: At one point these tags had hyphens instead of underscores.
+        # Since hyphens are interpreted as a subtraction operation in 
+        # Kusto Query Langauge (KQL) used by Azure Resource Graphs,
+        # these have been chnaged to underscores as a result.
+        return {
+            'ci_commit_sha1': sha,
+            'ci_manager': '',
+            unique_tag_key: tag_value}
+
+    def check_manager_exists(self, tag_value: str) -> bool:
+        inst = self.find_manager(tag_value)
+        return not (inst is None)
+    
+    def find_manager(self, tag_value: str):
+        instances = get_instances_with_filter([self.get_filter(tag_value), manager_filter])
+        if instances:
+            assert len(instances) == 1 # this must be called before any new instances are launched by workflow
+            return instances[0]
+        else:
+            return None
+
+    def find_all_workflow_instances(self, tag_value: str) -> List:
+        """ Grabs a list of all instance dicts sharing the CI workflow run's unique tag """
+        return get_instances_with_filter([self.get_filter(tag_value)])
+
+    def find_all_ci_instances(self) -> List:
+        """ Grabs a list of all instances across all CI using the CI unique tag key"""
+        all_ci_instances_filter = {'Name': 'tag:' + unique_tag_key, 'Values' : ['*']}
+        all_ci_instances = get_instances_with_filter([all_ci_instances_filter], allowed_states=['*'])
+        return all_ci_instances
+
+    def get_manager_ip(self, tag_value: str) -> str:
+        """
+        Looks up the AWS manager IP using the CI workflow run's unique tag
+        """
+        aws_manager = self.find_manager(tag_value)
+        if aws_manager is None:
+            raise Exception("No AWS manager instance running with tag matching the assigned workflow id\n")
+    
+        return aws_manager['PublicIpAddress']
```

Comment:
IIRC this value is only populated if the instance is properly started. Do you ensure the instance is properly running somewhere else? Or do you validate the IP is a valid IP somewhere else?

Reply:
I didn't change any of the logic for AWS, and I don't think we ever checked before. We usually print out the metadata information, including public ip, at launch, and it is usually populated there (although the print code does handle the case where it isn't). If you feel strongly about this, however, I can add an error check here. This case should never occur for Azure however.